File: src/css/values/rect.zig
const std = @import("std");
pub const css = @import("../css_parser.zig");
const Result = css.Result;
const Printer = css.Printer;
const PrintErr = css.PrintErr;
const DimensionPercentage = css.css_values.percentage.DimensionPercentage;
const LengthPercentage = css.css_values.length.LengthPercentage;
const LengthOrNumber = css.css_values.length.LengthOrNumber;
const CssColor = css.css_values.color.CssColor;

fn needsDeinit(comptime T: type) bool {
    return switch (T) {
        f32, i32, u32, []const u8 => false,
        LengthPercentage => true,
        LengthOrNumber => true,
        css.css_values.percentage.NumberOrPercentage => false,
        css.css_properties.border_image.BorderImageSideWidth => true,
        *const css.css_values.percentage.DimensionPercentage(css.css_values.length.LengthValue) => true,
        CssColor => true,
        css.css_properties.border.LineStyle => false,
        css.css_properties.border.BorderSideWidth => true,
        css.css_values.length.LengthPercentageOrAuto => true,
        else => @compileError("Don't know if " ++ @typeName(T) ++ " needs deinit. Please add it to this switch statement."),
    };
}

/// A generic value that represents a value for four sides of a box,
/// e.g. border-width, margin, padding, etc.
///
/// When serialized, as few components as possible are written when
/// there are duplicate values.
pub fn Rect(comptime T: type) type {
    const needs_deinit = needsDeinit(T);
    return struct {
        /// The top component.
        top: T,
        /// The right component.
        right: T,
        /// The bottom component.
        bottom: T,
        /// The left component.
        left: T,

        const This = @This();

        pub fn eql(this: *const This, other: *const This) bool {
            return css.generic.eql(T, &this.top, &other.top) and css.generic.eql(T, &this.right, &other.right) and css.generic.eql(T, &this.bottom, &other.bottom) and css.generic.eql(T, &this.left, &other.left);
        }

        pub fn deepClone(this: *const This, allocator: std.mem.Allocator) This {
            if (comptime needs_deinit or T == *const css.css_values.percentage.DimensionPercentage(css.css_values.length.LengthValue)) {
                return This{
                    .top = this.top.deepClone(allocator),
                    .right = this.right.deepClone(allocator),
                    .bottom = this.bottom.deepClone(allocator),
                    .left = this.left.deepClone(allocator),
                };
            }
            return This{
                .top = this.top,
                .right = this.right,
                .bottom = this.bottom,
                .left = this.left,
            };
        }

        pub fn all(val: T) This {
            return This{
                .top = val,
                .right = val,
                .bottom = val,
                .left = val,
            };
        }

        pub fn deinit(this: *const This, allocator: std.mem.Allocator) void {
            if (comptime needs_deinit) {
                this.top.deinit(allocator);
                this.right.deinit(allocator);
                this.bottom.deinit(allocator);
                this.left.deinit(allocator);
            }
        }

        pub fn parse(input: *css.Parser) Result(This) {
            return This.parseWith(input, valParse);
        }

        pub fn parseWith(input: *css.Parser, comptime parse_fn: anytype) Result(This) {
            const first = switch (parse_fn(input)) {
                .result => |vv| vv,
                .err => |e| return .{ .err = e },
            };
            const second = switch (input.tryParse(parse_fn, .{})) {
                .result => |v| v,
                // <first>
                .err => return .{ .result = This{ .top = first, .right = first, .bottom = first, .left = first } },
            };
            const third = switch (input.tryParse(parse_fn, .{})) {
                .result => |v| v,
                // <first> <second>
                .err => return .{ .result = This{ .top = first, .right = second, .bottom = first, .left = second } },
            };
            const fourth = switch (input.tryParse(parse_fn, .{})) {
                .result => |v| v,
                // <first> <second> <third>
                .err => return .{ .result = This{ .top = first, .right = second, .bottom = third, .left = second } },
            };
            // <first> <second> <third> <fourth>
            return .{ .result = This{ .top = first, .right = second, .bottom = third, .left = fourth } };
        }

        pub fn toCss(this: *const This, comptime W: type, dest: *Printer(W)) PrintErr!void {
            try css.generic.toCss(T, &this.top, W, dest);
            const same_vertical = css.generic.eql(T, &this.top, &this.bottom);
            const same_horizontal = css.generic.eql(T, &this.right, &this.left);
            if (same_vertical and same_horizontal and css.generic.eql(T, &this.top, &this.right)) {
                return;
            }
            try dest.writeStr(" ");
            try css.generic.toCss(T, &this.right, W, dest);
            if (same_vertical and same_horizontal) {
                return;
            }
            try dest.writeStr(" ");
            try css.generic.toCss(T, &this.bottom, W, dest);
            if (same_horizontal) {
                return;
            }
            try dest.writeStr(" ");
            try css.generic.toCss(T, &this.left, W, dest);
        }

        pub fn valParse(i: *css.Parser) Result(T) {
            return css.generic.parse(T, i);
        }

        pub fn isCompatible(this: *const @This(), browsers: css.targets.Browsers) bool {
            return this.top.isCompatible(browsers) and this.right.isCompatible(browsers) and this.bottom.isCompatible(browsers) and this.left.isCompatible(browsers);
        }
    };
}

File: src/css/values/rect.rs
use crate::css_parser as css;
use crate::css_parser::{CssResult as Result, PrintErr, Printer};
use crate::targets::Browsers;
use crate::values::protocol::{IsCompatible, Parse, ToCss};

// No `deinit` is needed: `T: Drop` on the four fields handles cleanup
// automatically (and arena-owned payloads in `bun_css` are bulk-freed by the
// bump, never per-value).

/// A generic value that represents a value for four sides of a box,
/// e.g. border-width, margin, padding, etc.
///
/// When serialized, as few components as possible are written when
/// there are duplicate values.
pub struct Rect<T> {
    /// The top component.
    pub top: T,
    /// The right component.
    pub right: T,
    /// The bottom component.
    pub bottom: T,
    /// The left component.
    pub left: T,
}

impl<T> Rect<T> {
    pub(crate) fn eql(&self, other: &Self) -> bool
    where
        T: PartialEq,
    {
        self.top == other.top
            && self.right == other.right
            && self.bottom == other.bottom
            && self.left == other.left
    }

    pub(crate) fn deep_clone(&self, _bump: &bun_alloc::Arena) -> Self
    where
        T: Clone,
    {
        // The `Clone` trait on `T` covers everything — the cheap-copy types
        // (`f32`, `NumberOrPercentage`, `LineStyle`) impl it as a bit copy.
        Self {
            top: self.top.clone(),
            right: self.right.clone(),
            bottom: self.bottom.clone(),
            left: self.left.clone(),
        }
    }

    pub(crate) fn all(val: T) -> Self
    where
        T: Clone,
    {
        Self {
            top: val.clone(),
            right: val.clone(),
            bottom: val.clone(),
            left: val,
        }
    }

    pub(crate) fn parse(input: &mut css::Parser) -> Result<Self>
    where
        T: Parse + Clone,
    {
        Self::parse_with(input, Self::val_parse)
    }

    pub(crate) fn parse_with<F>(input: &mut css::Parser, parse_fn: F) -> Result<Self>
    where
        F: Fn(&mut css::Parser) -> Result<T>,
        T: Clone,
    {
        let first = parse_fn(input)?;
        let second = match input.try_parse(&parse_fn) {
            Ok(v) => v,
            // <first>
            Err(_) => {
                return Ok(Self {
                    top: first.clone(),
                    right: first.clone(),
                    bottom: first.clone(),
                    left: first,
                });
            }
        };
        let third = match input.try_parse(&parse_fn) {
            Ok(v) => v,
            // <first> <second>
            Err(_) => {
                return Ok(Self {
                    top: first.clone(),
                    right: second.clone(),
                    bottom: first,
                    left: second,
                });
            }
        };
        let fourth = match input.try_parse(&parse_fn) {
            Ok(v) => v,
            // <first> <second> <third>
            Err(_) => {
                return Ok(Self {
                    top: first,
                    right: second.clone(),
                    bottom: third,
                    left: second,
                });
            }
        };
        // <first> <second> <third> <fourth>
        Ok(Self {
            top: first,
            right: second,
            bottom: third,
            left: fourth,
        })
    }

    pub(crate) fn to_css(&self, dest: &mut Printer) -> core::result::Result<(), PrintErr>
    where
        T: ToCss + PartialEq,
    {
        self.top.to_css(dest)?;
        let same_vertical = self.top == self.bottom;
        let same_horizontal = self.right == self.left;
        if same_vertical && same_horizontal && self.top == self.right {
            return Ok(());
        }
        dest.write_str(b" ")?;
        self.right.to_css(dest)?;
        if same_vertical && same_horizontal {
            return Ok(());
        }
        dest.write_str(b" ")?;
        self.bottom.to_css(dest)?;
        if same_horizontal {
            return Ok(());
        }
        dest.write_str(b" ")?;
        self.left.to_css(dest)
    }

    pub(crate) fn val_parse(i: &mut css::Parser) -> Result<T>
    where
        T: Parse,
    {
        T::parse(i)
    }
}

impl<T: IsCompatible> IsCompatible for Rect<T> {
    fn is_compatible(&self, browsers: &Browsers) -> bool {
        self.top.is_compatible(browsers)
            && self.right.is_compatible(browsers)
            && self.bottom.is_compatible(browsers)
            && self.left.is_compatible(browsers)
    }
}
